### PR TITLE
Add log browser

### PR DESCRIPTION
**Pull request by rafaellehmkuhl**
Log browser allow to download, delete and play autopilot logs (both `.tlog` and `.bin`).

Log browser uses `filebrowser` API to deal with the files and `UAV Logviewer` to play them.

image

Docker image available at `rafaellehmkuhl/companion-core/add-log-browser`

### CODE REVIEW

File: core/frontend/src/App.vue [hunk 1]
```diff
@@ -59,6 +59,12 @@
             </v-list-item-icon>
             <v-list-item-title>Main</v-list-item-title>
           </v-list-item>
+          <v-list-item to="/logs">
```
Comment (patrickelectric) [inline core/frontend/src/App.vue:62]: We should break this here with a line to divide in a "Vehicle" category and use a different icon other than home.
Maybe:
image

File: core/frontend/src/components/logs/LogManager.vue [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    are_logs_available(): boolean {
+      return this.available_logs.length !== 0
+    },
+  },
+  async mounted() {
+    await this.authenticateOnFileBrowser()
+    await this.fetchAvailableLogs()
+  },
+  methods: {
+    async authenticateOnFileBrowser(): Promise<void> {
```
Comment (patrickelectric) [inline core/frontend/src/components/logs/LogManager.vue:133]: We should break all this file handle logic API in a different class to be common for all our components.

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Besides that, this PR looks awesome

**Review by Williangalvani — COMMENTED** [review]
Nice!

I agree with @patrickelectric suggestion about the log icon.

Plus, should we filter for filetypes?
image

The size looks off too, I'd try to make it human-readable.

File: core/frontend/src/App.vue (hunk 1 above)
Comment (rafaellehmkuhl) [reply core/frontend/src/App.vue:62]: Didn't found this on Material Design Icons.

Think I will use the `file-multiple`:

image

**Comment by rafaellehmkuhl** [comment]
> Nice!
> 
> I agree with @patrickelectric suggestion about the log icon.
> 
> Plus, should we filter for filetypes?
> image
> 
> The size looks off too, I'd try to make it human-readable.

Cool.
Will filter for `.tlog` and `.bin` files.

**Review by patrickelectric — CHANGES_REQUESTED** [review]
- Disable download and delete buttons when no file is selected
    - This will download or remove / !
    - Add also a check to make sure that we are never going to use it with empty strings! 
- We should deal when one of the folder does not exist, like firmware/logs will not exist if navigator was never connected, this will make everything fail and no file will be available.
- The garbage icon should be red for DANGER and the play icon could be green
- The size of the files should be human friendly and not 0.0000085, maybe 85Bytes, 123Kb and etc, and not everything in MB.

**Comment by Williangalvani** [comment]
clicking the refresh button loses the sorting. I'm trying to get the latest log.
I would also consider just making it sorted by newest by default... thoughts?

**Comment by rafaellehmkuhl** [comment]
> clicking the refresh button loses the sorting. I'm trying to get the latest log.
> I would also consider just making it sorted by newest by default... thoughts?

Agree

**Comment by rafaellehmkuhl** [comment]
> clicking the refresh button loses the sorting. I'm trying to get the latest log.
> I would also consider just making it sorted by newest by default... thoughts?

Fixed sorting reset.

File: core/frontend/src/components/logs/LogManager.vue (hunk 2 above)
Comment (rafaellehmkuhl) [reply core/frontend/src/components/logs/LogManager.vue]: Done!

**Comment by rafaellehmkuhl** [comment]
Image updated.

File: core/frontend/src/libs/filebrowser.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  /* Download files (single or multiple). */
+  /**
+   * @param files - FilebrowserFile objects array
+  * */
+  downloadFiles(files: FilebrowserFile[]): void {
+    let url = ''
+    if (files.length === 1) {
+      url = this.singleFileRelativeURL(files[0])
```
Comment (Williangalvani) [inline core/frontend/src/libs/filebrowser.ts:143]: at a first glance, this looks unnecessary. isn't `multipleFilesRelativeUrl` able to handle a list with only one element?

Comment (Williangalvani) [reply core/frontend/src/libs/filebrowser.ts:143]: oh, it zips all the files together when it's multiple files? then I agree it is best to leave it as is.

File: core/frontend/src/App.vue [hunk 4]
```diff
@@ -90,6 +116,30 @@ export default Vue.extend({
...
+          },
+        ],
+      },
+    ],
   }),
 })
 </script>
+
+<style>
+.active_menu {
```
Comment (Williangalvani) [inline core/frontend/src/App.vue:142]: is this used? I couldn't find it.

**Review by Williangalvani — APPROVED** [review]
Looking great.

I have a couple of observations, but for me this is already good to go in. I'd be fine with tracking these in an issue or just opening a subsequent pr:

- Add "title" field to buttons (so their function shows up when hovering the mouse)
- The "page size" is lost when updating the table
- We should consider a confirmation dialog (or similar) when deleting files
- Human-readable sizes

File: core/frontend/src/libs/filebrowser.ts (hunk 3 above)
Comment (rafaellehmkuhl) [reply core/frontend/src/libs/filebrowser.ts:143]: > oh, it zips all the files together when it's multiple files? then I agree it is best to leave it as is.

Yup

File: core/frontend/src/App.vue (hunk 4 above)
Comment (rafaellehmkuhl) [reply core/frontend/src/App.vue]: No. Just updated without it.

**Comment by rafaellehmkuhl** [comment]
> Looking great.
> 
> I have a couple of observations, but for me this is already good to go in. I'd be fine with tracking these in an issue or just opening a subsequent pr:
> 
> * Add "title" field to buttons (so their function shows up when hovering the mouse)
> * The "page size" is lost when updating the table
> * We should consider a confirmation dialog (or similar) when deleting files
> * Human-readable sizes

Agree with all, but as you suggested I prefer to merge this as it is and open an issue with suggested improvements.

**Pull request merged**